chore: add a refactoring guide (#2413)
hashicorp/terraform-cdk: website/docs/cdktf/examples-and-guides/refactoring.mdx

---
page_title: Refactoring - CDK for Terraform
description: >-

>   As your CDKTF application grows in size and complexity, the need to move resources between stacks becomes more common. This guide will walk you through the process of refactoring your stacks.

Reviewer: ```suggestion
  Refactor stacks as your CDKTF application grows in size and complexity.
```

---

# Refactoring

> As your CDKTF application grows in size and complexity, the need to move resources between stacks becomes more common. This guide will walk you through the process of refactoring your stacks. In general you can change the names of stacks freely without an effect on the synthesized code. If you change the name of a resource this might lead to a re-creation of the resource, especially if you move the resource from one stack to another.

Reviewer: It might be good after this first sentence to give one real-world example that would make it a good idea to move resources around. Like "For example, you may need to z, y, z." Can we think of anything like that?

```suggestion
You may need to move resources between stacks as your CDK for Terraform (CDKTF) application grows in size and complexity. This guide demonstrates how to refactor your stacks. In general you can change the names of stacks freely without an effect on the synthesized code. If you change the name of a resource this might lead to a re-creation of the resource, especially if you move the resource from one stack to another.
```

> 

Reviewer 2: I think that's generally a hard question to answer in a document as short as this. 

That being said, I think we could be a bit more opinionated in our application structure that reduces the need for this kind of refactoring, but that's a conversation for a later time.

> 

Reviewer: okay that's fair. Even if it's just a few sentences, I think folks like guidance. But I trust your judgment that it may be too complex for this short space!

## Moving resources inside a stack

> Let's say you have a stack deployed with an aws `S3Bucket` resource and you want to give it another name:

Reviewer: ```suggestion
You may want to rename a stack deployed with an aws `S3Bucket` resource, like the following example.
```

> ## Moving resources inside a stack

Reviewer: @DanielMSchmidt all of these should be heading case (all major words capitalized). That's what the docs style guide says. I won't go through and make comments on each one, but would you be willing to go through and do that for all the titles?

Thank you :) 
```suggestion
## Moving Resources Inside a Stack
```

```typescript
import { Construct } from "constructs";
import { App, TerraformStack } from "cdktf";
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket";

class MyStack extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });

    // Old S3Bucket definition
    // new S3Bucket(this, "my-bucket", {
    //   bucketPrefix: "my-bucket",
    // });

    // New S3Bucket definition
    new S3Bucket(this, "new-bucket", {
      bucketPrefix: "my-bucket",
    });
  }
}

const app = new App();
new MyStack(app, "refactoring-example");
app.synth();
```

> When you run a `cdktf plan` you might see the following output:

Reviewer: We want to stay away from "see" to promote inclusion (per our style guide). I'm going to go through and suggest some alternatives, but just a note for future!

```suggestion
When you run a `cdktf plan`, CDKTF displays the following output:
```

```shell-session
refactoring-example    # aws_s3_bucket.my-bucket will be destroyed
                       # (because aws_s3_bucket.my-bucket is not in configuration)
                       - resource "aws_s3_bucket" "my-bucket" {
...
refactoring-example    # aws_s3_bucket.new-bucket (new-bucket) will be created
                       + resource "aws_s3_bucket" "new-bucket" {
```

> As you can see the old bucket is going to be destroyed and a new one is created to replace it. This can be harmful if there is already state in the resource (e.g. files in the bucket, data in the database, etc).

Reviewer: ```suggestion
Terraform plans to destroy the old bucket and create a new one to replace it. The replacement can be harmful if there is already state in the resource, such as files in the bucket or data in the database.
```

> To avoid the recreation of the resource you can use the [`terraform state mv`](/cli/commands/state/mv) command to move the state before the next apply.

Reviewer: ```suggestion
To avoid the recreation of the resource, use the [`terraform state mv`](/cli/commands/state/mv) command to move the state before the next apply.
```

```shell-session
# before the change, find the terraform name of the resource
# the name can be different than the name you put in the constructor of your construct,
# see the information below
$ cat cdktf.out/stacks/refactoring-example/cdk.tf.json | jq '.resource.aws_s3_bucket'
{
  "my-bucket": {
    "//": {
      "metadata": {
        "path": "refactoring-example/my-bucket",
        "uniqueId": "my-bucket"
      }
    },
    "bucket_prefix": "my-bucket"
  }
}

> # the id is the key of the object, in this case it is "new-bucket"

Reviewer 2: I feel like these should be outside the shell session. It's a multi-step process with extra explanation that can be broken apart for easier readability

# if you have multiple resources of the same type, you can use the path to find the right one

# Now change your code and synthesize it to get the new name of the resource
$ cdktf synth # we need to know the new id of the resource
$ cat cdktf.out/stacks/refactoring-example/cdk.tf.json | jq '.resource.aws_s3_bucket'
{
  "my-bucket": {
    "//": {
      "metadata": {
        "path": "refactoring-example/new-bucket",
        "uniqueId": "new-bucket"
      }
    },
    "bucket_prefix": "new-bucket"
  }
}

# Now that we know the old name is "my-bucket" and the new name is "new-bucket" we can move the state

> $ cd cdktf.out/stacks/refactoring-example

Reviewer 2: This part should be highlighted that we're running the rest of the steps in the generated stack folder

$ terraform init
# move the state
$ terraform state mv aws_s3_bucket.my-bucket aws_s3_bucket.new-bucket
Move "aws_s3_bucket.my-bucket" to "aws_s3_bucket.new-bucket"
Successfully moved 1 object(s).
# now you can apply the changes
$ cd ../../..
$ cdktf deploy
....
refactoring-example  aws_s3_bucket.new-bucket (new-bucket): Refreshing state... [id=my-bucket20221208132144348300000001]
refactoring-example  No changes. Your infrastructure matches the configuration.
...
```

### Why do I need to find the terraform name of the resource?

> If you use Constructs to organize your code you might have a level between the generated provider constructs and the `TerraformStack` construct for code organization purposes.
> This extra level adds a prefix to the name of the resource to ensure uniqueness. Read more about this in the [Constructs](/cdktf/concepts/constructs#scope) section.

Reviewer: ```suggestion
If you use Constructs to organize your code, you might have a level between the generated provider constructs and the `TerraformStack` construct to organize your code.

This extra level adds a prefix to the name of the resource to ensure uniqueness. Refer to [Constructs](/cdktf/concepts/constructs#scope) for more details.
```

## Moving resources from one stack to another

> Let's say you have a stack deployed with an aws `S3Bucket` resource and you want to move it to another stack:

Reviewer: ```suggestion
You may want to move a resource from one stack to another, like the following example aws `S3Bucket` resource.
```

```typescript
import { Construct } from "constructs";
import { App, TerraformStack } from "cdktf";
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket";

class MyFirstStack extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });

    // Old Definition
    // new S3Bucket(this, "my-bucket", {
    //   bucketPrefix: "my-bucket",
    // });
  }
}

class MySecondStack extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });

    // New Definition
    new S3Bucket(this, "my-other-bucket", {
      bucketPrefix: "my-bucket",
    });
  }
}

const app = new App();
new MyFirstStack(app, "refactoring-example-stack-move-1");
new MySecondStack(app, "refactoring-example-stack-move-2");
app.synth();
```

> When we run a `cdktf deploy '*'` we will see the following output:

Reviewer: ```suggestion
When you run a `cdktf deploy '*'`, the CLI displays the following output.
```

```shell-session
...
refactoring-example-stack-move-1    # aws_s3_bucket.my-bucket will be destroyed
                                    # (because aws_s3_bucket.my-bucket is not in configuration)
                                    - resource "aws_s3_bucket" "my-bucket" {
...
refactoring-example-stack-move-2    # aws_s3_bucket.my-other-bucket (my-other-bucket) will be created
...
```

> If we want to preserve the state of the resource we need to move it from one stack to another. For this we need to use [`terraform import`](/cli/commands/import) and [`terraform state rm`](/cli/commands/state/rm).

Reviewer: ```suggestion
To preserve the state of the resource, you must move it from one stack to another. You can use  [`terraform import`](/cli/commands/import) and [`terraform state rm`](/cli/commands/state/rm) to move the resource's state.
```

```shell-session
# First we need to find the terraform name of the resource in the old stack
# Run this before removing the resource from the old stack
$ cat cdktf.out/stacks/refactoring-example-stack-move-1/cdk.tf.json | jq '.resource.aws_s3_bucket'
{
  "my-bucket": {
    "//": {
      "metadata": {
        "path": "refactoring-example-stack-move-1/my-bucket",
        "uniqueId": "my-bucket"
      }
    },
    "bucket_prefix": "my-bucket"
  }
}

$ cdktf synth # To ensure our synthesized stack is up to date

> # Now we need the name of the resource in the new stack to understand

Reviewer 2: Similar to above, this can be broken apart to multiple code snippets + text

Reviewer: +1 to this suggestion :)

$ cd cdktf.out/stacks/refactoring-example-stack-move-1/
$ terraform init
$ terraform state show aws_s3_bucket.my-bucket
# aws_s3_bucket.my-bucket:
resource "aws_s3_bucket" "my-bucket" {
    arn                         = "arn:aws:s3:::my-bucket20221208141249058600000001"
    bucket                      = "my-bucket20221208141249058600000001"
    bucket_domain_name          = "my-bucket20221208141249058600000001.s3.amazonaws.com"
    bucket_prefix               = "my-bucket"
    bucket_regional_domain_name = "my-bucket20221208141249058600000001.s3.amazonaws.com"
    force_destroy               = false
    hosted_zone_id              = "Z3AQBSTGFYJSTF"
    id                          = "my-bucket20221208141249058600000001"
    object_lock_enabled         = false
    region                      = "us-east-1"
    request_payer               = "BucketOwner"
    tags_all                    = {}

    grant {
        id          = "554912fda2704333d162d216be50aefb05562e0bf1709997f1d9417cf46087d5"
        permissions = [
            "FULL_CONTROL",
        ]
        type        = "CanonicalUser"
    }

    versioning {
        enabled    = false
        mfa_delete = false
    }
}

> # Now that we have the id ("my-bucket20221208141249058600000001") we can import it

Reviewer 2: Hmm, from what I recall, import for different resources use different kinds of IDs (ID vs. ARN vs. something else)? If that's true, we should call it out, but if not, ignore this comment.

Reviewer 3: Maybe copy parts of these Terraform docs (or link to them?):

> ID is dependent on the resource type being imported. For example, for AWS EC2 instances it is the instance ID (i-abcd1234) but for AWS Route53 zones it is the zone ID (Z12ABC4UGMOZ2N). Please reference the provider documentation for details on the ID format. If you're unsure, feel free to just try an ID. If the ID is invalid, you'll just receive an error message.

https://developer.hashicorp.com/terraform/cli/commands/import#usage

Author: I copied the start and linked to the rest 👍
